### ISSUE TITLE
Feature request: continuous view

### ISSUE THREAD

**Issue by oscarfv** [politza/pdf-tools#27]
Is it feasible to implement viewing the top of the next page when you reach the end of the current page? While reading e-books and scrolling down with `Space` as you advance it is unpleasant that when the bottom of the page is reached the line that was previously at the bottom of the screen now it can be located anywhere depending on the zoom level etc. Other pdf viewers solve this by showing the top of the next page below the current one and thus always advancing the same amount of space on each scroll operation.

Ditto for scrolling up.

**Comment by politza** [comment]
We could have a `fixed-scroll-mode`.

Displaying more than one page/image per window is not going to happen
anytime soon.

**Comment by oscarfv** [comment]
How would work that `fixed-scroll-mode`?

If Emacs could do the same amount of scrolling every time `Space` is hitted, it would be good enough for now. Maybe adding empty lines after the image?

**Comment by politza** [comment]
>> How would work that fixed-scroll-mode?

Like you described.

**Comment by titaniumbones** [comment]
Right now the lack of a continuous mode is the only thing that I miss from other viewers. From what you say, this is unlikely to work -- I guess because the buffer is only capable of displaying one image at a time?  And is that a fundamental constraint?  

In any case I find the jump fro mthe bottom of one page to the top of the next hard to follow.  It's too bad because otherwise pdf-tools is already way better than any other viewer!

**Comment by politza** [comment]
Matt Price <notifications@github.com> writes:

> [...] the buffer is only capable of displaying one image at a time?
> And is that a fundamental constraint? [...]

No, but there are some difficulties to overcome. I already have an idea
how, but haven't found the energy to go ahead.  I could elaborate, If
you care.

**Comment by thisirs** [comment]
I'd love the see this feature implemented... This is actually the only thing that keeps me from switching to pdf-tools (which is great btw ;))

**Comment by antoine-levitt** [comment]
+1 on the comments, this package is great and I'd love to switch to it, but the lack of continuous scrolling is what's keeping me back.

**Comment by clubfed** [comment]
+1

**Comment by Compro-Prasad** [comment]
I think that it would be better if you could elaborate so that someone could get how to solve this problem and submit a merge request.

**Comment by guilhermemtr** [comment]
I also feel that it would be a great new feature to pdf-tools (I actually subscribed to this issue waiting for updates on this feature).

This is what I think is meant by the feature request (illustrated with images, using evince):
Image 1:
![screenshot from 2017-07-21 16-54-05](https://user-images.githubusercontent.com/5527193/28471572-7f5a7664-6e35-11e7-9e2e-b22a8d9a1893.png)

Image 2:
![screenshot from 2017-07-21 16-53-48](https://user-images.githubusercontent.com/5527193/28471573-7f5b3bbc-6e35-11e7-9066-9e405c2898fd.png)

Basically, it would be great to be able to scroll the whole document (instead of only having the page by page display), and have the pages appearing as if they were only one very long page (but keeping the delimiters, as in the images).

**Comment by politza** [comment]
Generally speaking, there are two issues:

1. Large parts othe software (including image-mode) assume a one-to-one correspondence between a displayed page and its window.
2. Emacs is not easily convinced to scroll an image, such that its display starts in the middle of it,  in case its completly fits into the window.

**Comment by Compro-Prasad** [comment]
Can't we display a limited number of images vertically(max 4 pages) in the same buffer and scroll it normally. When we hit the next page or the next next page we can remove images above and add new ones below. I mean it may be something like `(save-excursion (redisplay-images))` .

**Comment by Compro-Prasad** [comment]
Spacemacs accomplishes a scrolling behaviour by remapping the keys but with no continuous effect.

**Comment by guilhermemtr** [comment]
If I programmed in el I would do the request, but I don't.
Maybe, here is a way to do it:

Use 5 images [0,1,2,3,4].
When the user scrolls out of image[2], start loading image[0] or [4] (depending if it's a scroll up or down).
When the user is reaching the bottom half of [3] (or top-half of [1]): shift the image array down (or up)*, load the new image in [4] (or [0]) and make the pdf shift the view to make a smoother transition (and hopefully seemless)

*e.g. [img1,img2,img3,img4,img5] -> [img2,img3,img4,img5,img??]
Is this doable?
I'm not aware of the implementation details so it might be completely out.

**Comment by hisnawi** [comment]
Any updates on this?

**Comment by pprw** [comment]
+1 on this feature request. I personally want also to see double page mode on pdf-tools (see https://github.com/politza/pdf-tools/issues/303).

**Comment by Compro-Prasad** [comment]
I think we should have a feature request in Emacs for an infinite background scroll mode. Which would help in solving this issue.

**Comment by narendraj9** [comment]
@politza How about merging images of two pages together into a single image with the line break in between?

**Comment by tsktsktsk64** [comment]
+1 for this feature.

**Comment by bricewge** [comment]
With the release of emacs 26.1 there is a new mode called `pixel-scroll-mode` that could help us having continus view in `pdf-tools`; I didn't tried it though.

**Comment by Compro-Prasad** [comment]
`pixel-scroll-mode` isn't production ready yet. It lags seriously on my machine. I am running Arch Linux.

**Comment by FrancescoCadei** [comment]
+1 with this feature i could free myself from any other pdf-viewer.

**Comment by Emrys-Merlin** [comment]
+1 from me, too :-)

**Comment by braham-snyder** [comment]
Though I too would love to see this feature, "+1"-style posts send every subscriber a notification -- I (and I suspect many other subscribers) would greatly appreciate it if everyone instead gave the OP a thumbs up (by clicking the smiley face in its top right corner).

(For those who haven't ran into it: GitHub's issue tracker can sort issues by the numbers of reactions they've each garnered -- ideally, IMO, we'd have explicit issue voting with votes displayed next to the number of comments as in https://github.com/isaacs/github/issues/9, but unfortunately that didn't look like it was going to happen last I checked.)

**Comment by shader** [comment]
Would it be reasonable to use `convert -append` (or whatever the library equivalent is) to vertically attach the images from several pages together? Then we could put 2-3 pages into one PNG, and it wouldn't matter whether or not emacs natively supported displaying multiple images. I know that emacs already uses image-magick in other places. I don't know how much that would mess up the interactive parts of pdf-tools though.

**Comment by politza** [comment]
In my opinion, there is no good way to support multiple images per window without rewriting large parts of this package.  I've started something like this, though I currently don't have much time to spend on it.

And yes, what you're proposing  would mess up almost everything.

Also, it's not Emacs that is the problem, but the assumptions made by this package.

**Comment by oscarfv** [comment]
One of the main conveniences of continuous scroll mode is to visually keep track of the point you are looking at when the image scrolls up/down. My recent experience with ebook readers is that an animated scroll helps a lot on this regard. The animation needs to be "smooth" and "progressive" so that the eyes effortlessly keep track of the point of interest.

About a year or two ago I looked at `pixel-scroll-mode` and found it inadequate, if not utterly broken, for scrolling images. Maybe now things are better. Also, recent Emacs development introduced improvements on image handling (rotation and zooming, mostly, but it would be interesting to inquire if the improvements can be extended to pixel-based positioning and glitch-free fast repainting).

**Comment by politza** [comment]
At least scrolling the window such that an image is only partially displayed if it could be completely displayed otherwise seems to be possible with recent Emacs versions.

**Comment by forgetmenotz** [comment]
Just wondering if there is any possibility of ever seeing this introduced or if its abandoned at the moment? One page at a time is too clunky.

I compiled Koreader for arch recently which allows continuous scrolling, and with some hacks, key based half page paging, search and maybe highlighting. The reason being, I get visual defects in Zathura. But of course the power of Emacs is still missing. And zoom is not working for me either in pdf-tools.

**Comment by ThibautVerron** [comment]
Off the top of my head, you could also try the pdf viewer in the [emacs application framework](https://github.com/manateelazycat/emacs-application-framework).

**Comment by ThibautVerron** [comment]
> In my opinion, there is no good way to support multiple images per window without rewriting large parts of this package.  I've started something like this, though I currently don't have much time to spend on it. And yes, what you're proposing  would mess up almost everything. Also, it's not Emacs that is the problem, but the assumptions made by this package.

Is it really "large parts of the package", or just essentially most of `pdf-util.el`? Or is it like a large part of the code makes assumptions beyond the output of the tools defined in `pdf-util` and those would need to be found/examined one by one?

**Comment by tmalsburg** [comment]
> Generally speaking, there are two issues:
> - Large parts of the software (including image-mode) assume a one-to-one correspondence between a displayed page and its window.
> - Emacs is not easily convinced to scroll an image, such that its display starts in the middle of it, in case would completely fits into the window.

This would be a bit of a hack but perhaps it is good enough while successfully working around
Emacs’ limitations?:
- Display one image that shows three pages.  The current page, the previous, and the next.
- When the user scrolls past the boundary between the current page and the next, replace the image with one where the next page becomes the middle one.
- Same for scrolling up.

Could something like that work?

**Comment by dalanicolai** [comment]
I have created a dirty hack [in this file](https://github.com/dalanicolai/dala-emacs-lisp/blob/master/pdf-continuous-scroll.el).
It implements continuous scroll by creating a second buffer when needed and resizing it while scrolling.
I might be useful for people who "desperately" ;) require continuous scrolling behavior.
It looks even better when you remove the modeline (which can be achieved by uncommenting the hook at the beginning of the file), but unfortunately it seems to create some problems for which I did not find the cause yet.
Also it contains a function to toggle between the normal continuous view behavior and the hacked behavior.

You can use it by downloading the file and add something similar to 
```
  (setq pdf-view-continuous 'hack)
  (eval-after-load "pdf-view"
    '(load "~/git/dala-emacs-lisp/pdf-continuous-scroll.el"))
```
in your user-config.

**Comment by dalanicolai** [comment]
https://github.com/dalanicolai/pdf-continuous-scroll-mode.el

**Comment by dalanicolai** [comment]
For the ones still interested... I have just created a (not very, but still) simple sketch to obtain "real" continuous scrolling (i.e. in a single buffer). You can check it out here. It would be great if others would also like to hack on it. With pdf-tools, and docview as an example, together with the very simple to use python-epc and the amazing pymupdf documentation, a lot should be possible. On the other hand I understand that, like for myself, this will not have priority. But it could be fun and educational. I started this for fun and to show a proof of concept.

**Comment by dalanicolai** [comment]
Another project would be to replace `epdfinfo` with the `pymupdf` plus `python-epc` solution in `pdf-tools`. It would provide possibilities to add more styles of annotations like in [pymupdf-mode](https://github.com/dalanicolai/pymupdf-mode.el) and support for editing forms.

**Comment by oscarfv** [comment]
> Another project would be to replace `epdfinfo` with the `pymupdf` plus `python-epc` solution in `pdf-tools`. It would provide possibilities to add more styles of annotations like in [pymupdf-mode](https://github.com/dalanicolai/pymupdf-mode.el) and support for editing forms.

For some people adding Python as a dependency is not trivial.

**Comment by emacsomancer** [comment]
@dalan

> For the ones still interested... I have just created a (not very, but still) simple sketch to obtain "real" continuous scrolling (i.e. in a single buffer). You can check it out here. It would be great if others would also like to hack on it. With `pdf-tools`, and `docview` as an example, together with the very simple to use `python-epc` and the amazing `pymupdf` documentation, a lot should be possible. On the other hand, I understand that, like for myself, this will not have priority. But it could be fun and educational. I started this for fun and to show a proof of concept.

(missing link: https://github.com/dalanicolai/pdf-continuous-scroll-mode.el ? )

**Comment by dalanicolai** [comment]
@emacsomancer Thanks for pointing that out, I have added the link to https://github.com/dalanicolai/vimura-epc now

**Comment by bob333** [comment]
Thanks for this package. The following code will help to install this package since its not available in melpa, It does by first installing quelpa and then using it to install this package.

```
(use-package quelpa
  :ensure t)

(use-package pdf-tools
  :ensure t
  :after (quelpa)
  :config
  (pdf-tools-install t)
  (quelpa '(pdf-continuous-scroll-mode
          :fetcher github
          :repo "dalanicolai/pdf-continuous-scroll-mode.el"))

  (add-hook 'pdf-view-mode-hook 'pdf-continuous-scroll-mode))
```

**Comment by dalanicolai** [comment]
For users who end up here when looking for continuous-view, the [image-roll package](https://github.com/dalanicolai/image-roll.el) provides such functionality.
